### PR TITLE
TreeStatus api for EditableTree/UntypedTree

### PR DESCRIPTION
**Pull request by daesunp**
## Description

This PR does the following:
1. Adds a `TreeStatus` to `EditableTree`/`EditableField` and `UntypedTree`/`UntypedField` which allows users to learn about a given tree/field has been removed vs deleted.
2. Renames the `delete` apis to `remove` apis, to better reflect what is actually happening.

### CODE REVIEW

File: api-report/tree2.api.md [hunk 1]
```diff
@@ -1917,6 +1920,16 @@ type TreeSchemaSpecification = [
...
 
+// @alpha
+export enum TreeStatus {
+    // (undocumented)
+    Deleted = 2,
+    // (undocumented)
+    InDocument = 0,
+    // (undocumented)
+    Removed = 1
+}
```
Comment (CraigMacomber) [inline api-report/tree2.api.md:1931]: put the doc comment for each status on the enum member itself. This will improve intelisense fo them, and will make them not show up as undocumented here.

File: experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts [hunk 2]
```diff
@@ -392,11 +394,23 @@ export class FieldProxyTarget extends ProxyTarget<FieldAnchor> implements Editab
...
 
+	public getTreeStatus(): TreeStatus {
+		const path = this.cursor.getFieldPath().parent;
+		if (path === undefined) {
+			return TreeStatus.Deleted;
+		}
+		/**
+		 * TODO: This is a slow initial implementation which traverses the entire path up to the root of the tree.
+		 * This should eventually be optimized.
+		 */
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts:405]: There is no reason to use TsDoc comments inside on a a line of code that is not declaring something. Just use // comments for cases like this.

File: experimental/dds/tree2/src/feature-libraries/editable-tree/utilities.ts [hunk 3]
```diff
@@ -99,3 +99,23 @@ export function getStableNodeKey(
 		return brand(id);
 	}
 }
+
+/**
+ * Checks whether or not a given path is parented under the root field.
+ * @param path - the path you want to check.
+ * @returns a boolean on whether or not the provided path is parented under the root field.
+ */
+export function isParentedUnderRootField(path: UpPath): boolean {
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/feature-libraries/editable-tree/utilities.ts:108]: We can make this logic more reusable with a little refactoring. Turn this into "getDetachedSequenceContainingPath" and it will be usable in more cases.

Comment (CraigMacomber) [inline experimental/dds/tree2/src/feature-libraries/editable-tree/utilities.ts:108]: This function should be easy to unit test, so add some tests for it to this files tests file.

File: experimental/dds/tree2/src/feature-libraries/editable-tree/utilities.ts [hunk 4]
```diff
@@ -99,3 +99,23 @@ export function getStableNodeKey(
...
+	let currentPath = path;
+	while (currentPath.parentField !== undefined) {
+		if (currentPath.parentField === rootFieldKey) {
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/feature-libraries/editable-tree/utilities.ts:111]: This isn't the correct way to do this check. A path that has this key somewhere along the path (but not as the root) should not pass this check.

I'd suggest adding a regression test for this case, but if you do the suggested change to getDetachedSequenceContainingPath I mentioned above, its impossible to have (or test for) this bug.

File: experimental/dds/tree2/src/test/feature-libraries/editable-tree/editableTree.editing.spec.ts [hunk 5]
```diff
@@ -528,7 +528,7 @@ describe("editable-tree: editing", () => {
 			});
 
 			// delete method
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/test/feature-libraries/editable-tree/editableTree.editing.spec.ts:530]: thes comments (here and below) also should be updated

File: api-report/tree2.api.md (hunk 1 above)
Comment (daesunp) [reply api-report/tree2.api.md]: Updated doc comment for each status on enum

File: experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts (hunk 2 above)
Comment (daesunp) [reply experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts]: removed TsDoc comment and used //

File: experimental/dds/tree2/src/feature-libraries/editable-tree/utilities.ts (hunk 3 above)
Comment (daesunp) [reply experimental/dds/tree2/src/feature-libraries/editable-tree/utilities.ts]: Refactored to `getDetachedSequenceContainingPath`

Comment (daesunp) [reply experimental/dds/tree2/src/feature-libraries/editable-tree/utilities.ts]: Added unit tests for `getDetachedSequenceContainingPath`

File: experimental/dds/tree2/src/feature-libraries/editable-tree/utilities.ts (hunk 4 above)
Comment (daesunp) [reply experimental/dds/tree2/src/feature-libraries/editable-tree/utilities.ts]: Refactored to `getDetachedFieldContainingPath` as suggested above

File: experimental/dds/tree2/src/test/feature-libraries/editable-tree/editableTree.editing.spec.ts (hunk 5 above)
Comment (daesunp) [reply experimental/dds/tree2/src/test/feature-libraries/editable-tree/editableTree.editing.spec.ts]: updated comments to "remove" instead of "delete"

File: experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts [hunk 6]
```diff
@@ -392,11 +395,23 @@ export class FieldProxyTarget extends ProxyTarget<FieldAnchor> implements Editab
 		);
 	}
 
+	public getTreeStatus(): TreeStatus {
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts:398]: since this is a method on a tree (and it also clarified in the return type), I think `tree` in its name is redundant, so we can shorten this to "getStatus()".

File: experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts [hunk 7]
```diff
@@ -392,11 +395,23 @@ export class FieldProxyTarget extends ProxyTarget<FieldAnchor> implements Editab
...
+		const path = this.cursor.getFieldPath().parent;
+		if (path === undefined) {
+			return TreeStatus.Deleted;
+		}
+		// TODO: This is a slow initial implementation which traverses the entire path up to the root of the tree.
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts:403]: We arn't consistent about it, and don't document it anywhere, but one convention I like from my previous team it to make performance related TODOs more searchable like this:

```suggestion
		// TODO:Performance: This is a slow initial implementation which traverses the entire path up to the root of the tree.
```

Comment (CraigMacomber) [reply experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts:403]: Also, I don't like code which contains TODO's being duplicated: better to have one call the other or have them both call a shared helper which contains the TODO to reduce TODOs in the code and make them easier to implement without missing any ussages.

Comment (CraigMacomber) [reply experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts:403]: Also factoring this policy out into its own function will make testing the optimized version against it easier (the simple version can be moved to a test utility and used to test against if needed)

File: experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts [hunk 8]
```diff
@@ -392,11 +395,23 @@ export class FieldProxyTarget extends ProxyTarget<FieldAnchor> implements Editab
 		);
 	}
 
+	public getTreeStatus(): TreeStatus {
+		const path = this.cursor.getFieldPath().parent;
+		if (path === undefined) {
+			return TreeStatus.Deleted;
+		}
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts:402]: I don't think this is the right approach to detecting deleted for a few reasons:
1. if the editable tree node is deleted, accessing its cursor is invalid and ideally would error.
2. This will return true for any root field (the special root or any other detached field)
3. If using the path information, the same data from the anchor instead of cursor will be faster.

For the deleted checking, I recommend adding a method to ProxyTarget, "isFreed" which returns true iff free has been called. TO avoid having to tract ans extra property for that, it can compute it by checking this.lazyCursor's status.

Comment (CraigMacomber) [reply experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts:402]: Also, given that your implementation aparently has a bug (incorrectly reports deleted for root fields, and doesn't report deleted correctly), I'm guessing it could use some more tests to ensure it is and remains correct in these aspects.

File: experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts (hunk 7 above)
Comment (daesunp) [reply experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts]: updated TODO comment and created helper function to remove redundant TODO comment

File: experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts (hunk 6 above)
Comment (daesunp) [reply experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts]: updated method name to `getStatus`

File: experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts (hunk 8 above)
Comment (daesunp) [reply experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts]: Added `isFreed` method to `ProxyTarget` and updated logic for `TreeStatus.Delete`. Added unit tests to check `treeStatus` method

File: experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts [hunk 9]
```diff
@@ -392,11 +396,29 @@ export class FieldProxyTarget extends ProxyTarget<FieldAnchor> implements Editab
...
+		// If the parentAnchor is undefined it is a detached field.
+		if (parentAnchor === undefined) {
+			return keyAsDetachedField(fieldAnchor.fieldKey) === rootField
+				? TreeStatus.InDocument
+				: TreeStatus.Removed;
+		}
+		const parentAnchorNode = this.context.forest.anchors.locate(parentAnchor);
+		// As the "parentAnchor === undefined" case is handled above, parentAnchorNode should exist.
+		assert(parentAnchorNode !== undefined, "parentAnchorNode must exist.");
+		return treeStatusFromPath(parentAnchorNode);
```
Comment (daesunp) [inline experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts:414]: I changed the logic using anchors instead of cursors, but still couldn't get around using an assert statement as the `locate()` method can still return undefined. Since we already checked that `parentAnchor` is not undefined (from the if statement above), would it be safe to keep this assert here?

File: experimental/dds/tree2/src/test/feature-libraries/editable-tree/editableTree.editing.spec.ts [hunk 10]
```diff
@@ -662,9 +664,80 @@ describe("editable-tree: editing", () => {
...
+				assert(isEditableField(field));
+				const statusBeforeRemove = field.treeStatus();
+				assert.equal(statusBeforeRemove, TreeStatus.InDocument);
+
+				const rootField = view.context.root;
+				rootField.remove();
+
+				const statusAfterRemove = field.treeStatus();
+				assert.equal(statusAfterRemove, TreeStatus.Deleted);
 			});
```
Comment (daesunp) [inline experimental/dds/tree2/src/test/feature-libraries/editable-tree/editableTree.editing.spec.ts:741]: It seems that when I try to call the `treeStatus()` method on a removed field, it asserts error `0x3c4 /* It is invalid to access an EditableTree node which no longer exists */`. Is there a good way to get around this?

File: experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts (hunk 9 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts:414]: You shouldn't be doing `const parentAnchor = fieldAnchor.parent;` nor asserting `fieldAnchor.parent !== undefined`.

That case is valid, and you should handle it with an if statement, not crash when its the case. Your goal here is to find the key of the root field. When your path does not have a parent, that means you are at the root, and need to check that field key. By throwing away that field key by only using the parent of the fieldPath from the anchor, you make this impossible.

Comment (CraigMacomber) [reply experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts:414]: This is fine: you accurately describe why the assert should always pass (and I agree its a valid assumption). The fact that there isn't a good way to avoid this assert is a hint that maybe we could provide better APIs since it means we have assumptions that arn't captured in the types. For example if we had:

```typescript
export function locateFieldAnchor(locator: AnchorLocator, anchor: FieldAnchor): FieldUpPath<AnchorNode> { ... }
```

You could call locateFieldAnchor at the beginning, then only check for undefined parent in once place (in the returned FieldUpPath) instead of in both the if and assert.

If you write that function (after https://github.com/microsoft/FluidFramework/pull/16950 goes in), it will also make testing a little easier since slightly more logic is moved into a simpler to test place.

That said, this code is fine as is, so feel free to leave if you want.

File: experimental/dds/tree2/src/test/feature-libraries/editable-tree/editableTree.editing.spec.ts [hunk 11]
```diff
@@ -662,9 +664,80 @@ describe("editable-tree: editing", () => {
...
+					initialTree: { foo: [], foo2: [] },
+				});
+				const rootNode = view.root;
+				assert(isEditableTree(rootNode));
+				const rootNodeStatus = rootNode[treeStatus]();
+
+				assert.equal(rootNodeStatus, TreeStatus.InDocument);
+			});
+
+			it("removed node returns TreeStatus.Deleted", () => {
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/test/feature-libraries/editable-tree/editableTree.editing.spec.ts:689]: would be good to include a comment on or in this test for why it does not return "removed" yet.

File: experimental/dds/tree2/src/test/feature-libraries/editable-tree/editableTree.editing.spec.ts [hunk 12]
```diff
@@ -662,9 +664,80 @@ describe("editable-tree: editing", () => {
...
+				const rootNodeStatus = rootField.treeStatus();
+
+				assert.equal(rootNodeStatus, TreeStatus.InDocument);
+			});
+
+			it("removed field returns TreeStatus.Deleted", () => {
+				const view = createSharedTreeView().schematize({
+					schema: getTestSchema(FieldKinds.sequence),
+					allowedSchemaModifications: AllowedUpdateType.None,
+					initialTree: { foo: ["foo"], foo2: [] },
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/test/feature-libraries/editable-tree/editableTree.editing.spec.ts:727]: In both the node and field tests, you include an extra child, but don't do anything with it.

You should probably extend the tests so that they ensure the contents of deleted/removed trees are also considered deleted/removed.

You also only test the inDocument case at the root. either extends (and rename) those tests to check non-root cases or add addationl tests for non-root fields and nodes.

While there isn't much special logic for non-root nodes (so maybe just renaming the node test to "node returns TreeStatus.InDocument" is enough for it, as "root" nodes arn't that special, root fields had different code handleing them then non-root fields, so I think they deserve a could extra lines in the test or a separate test.

File: experimental/dds/tree2/src/test/feature-libraries/editable-tree/editableTree.editing.spec.ts (hunk 10 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/test/feature-libraries/editable-tree/editableTree.editing.spec.ts:741]: Look at the call stack and see why its erroring. Its possible you will need to avoid accessing something on a deleated tree thats invalid and/or move where the check for deleted trees happens.

File: experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts [hunk 13]
```diff
@@ -392,11 +396,33 @@ export class FieldProxyTarget extends ProxyTarget<FieldAnchor> implements Editab
...
+				: TreeStatus.Removed;
+		}
+		const parentAnchorNode = this.context.forest.anchors.locate(parentAnchor);
+
+		// The parentAnchorNode can be undefined in the case that its corresponding PathNode !== Status.Alive.
+		// This means that the parentAnchor points to an invalid node.
+		if (parentAnchorNode === undefined) {
+			return TreeStatus.Deleted;
+		}
+		return treeStatusFromPath(parentAnchorNode);
```
Comment (daesunp) [inline experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts:418]: The `anchors.locate` method seems to return `undefined` when its corresponding `PathNode` does not have the status `Status.Alive`. I updated the assert statement to use an if statement instead, and return `TreeStatus.Delete` in this case, but I am not entirely sure if this is correct.

Based on the documentation there seems to be 2 other statuses `Status.Disposed` and `Status.Dangling`. `Status.Dangling` seems to correspond to a deleted node. But does `Status.Disposed` mean that it's removed (but invalid)? If that's the case, we may need to distinguish between the two cases, but there seems to be no good way to get the `PathNode` status.

Should an additional method along the lines of `getPathNodeStatus(anchor:Anchor)` be created in `AnchorSet` so that we could make this distinction?

File: experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts (hunk 9 above)
Comment (daesunp) [reply experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts:418]: Updated assert to if statement due to the following reason:

The anchors.locate method seems to return undefined when its corresponding PathNode does not have the status Status.Alive. I updated the assert statement to use an if statement instead, and return TreeStatus.Delete in this case, but I am not entirely sure if this is correct.

Based on the documentation there seems to be 2 other statuses Status.Disposed and Status.Dangling. Status.Dangling seems to correspond to a deleted node. But does Status.Disposed mean that it's removed (but invalid)? If that's the case, we may need to distinguish between the two cases, but there seems to be no good way to get the PathNode status.

Should an additional method along the lines of getPathNodeStatus(anchor:Anchor) be created in AnchorSet so that we could make this distinction?

File: experimental/dds/tree2/src/test/feature-libraries/editable-tree/editableTree.editing.spec.ts (hunk 12 above)
Comment (daesunp) [reply experimental/dds/tree2/src/test/feature-libraries/editable-tree/editableTree.editing.spec.ts:751]: Updated test cases

File: experimental/dds/tree2/src/test/feature-libraries/editable-tree/editableTree.editing.spec.ts (hunk 11 above)
Comment (daesunp) [reply experimental/dds/tree2/src/test/feature-libraries/editable-tree/editableTree.editing.spec.ts]: Added comment to test cases for clarification.

File: experimental/dds/tree2/src/test/feature-libraries/editable-tree/editableTree.editing.spec.ts (hunk 10 above)
Comment (daesunp) [reply experimental/dds/tree2/src/test/feature-libraries/editable-tree/editableTree.editing.spec.ts:775]: Changed the logic order in `get` of `FieldProxyTarget`

File: experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts (hunk 9 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts:418]: You should never see any status other than alive since if you have an ProxyTarget that has not been freed: the ProxyTarget holds onto an anchor, which ensures the AnchorNode doesn't get disposed, and if the AnchorNode does become dangling, it will have triggered its "afterDelete" event, which frees the ProxyTarget.

File: experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts (hunk 13 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/feature-libraries/editable-tree/editableField.ts:418]: It should be impossible for well-formed code to try and look up a disposed AnchorNode, since the only way to look up an AnchorNode is from an anchor, which is a ref count that keeps the AnchorNode from being disposed.

Also most usages (including editable tree) shouldn't see the dangling state either, since they should really free their anchor when they see a delete event since theres no real point to holding onto an anchor after that point.

**Review by CraigMacomber — APPROVED** [review]
